FO GUI Sorting system implemented, Revive System Modification

NeverendingLearner opened the pull request:
Gives the user the ability to sort by item categories.

NeverendingLearner committed:
Sorting system implemented

Gives the user the ability to sort by item categories.

NeverendingLearner committed:
Added pictures to display

The pictures consist of:
- A big picture to the right of the weapons list
- A small picture right next to the weapons individual name

10xJSChad commented on A3A/addons/core/functions/FactoryOverhaul/fn_productionGUIControls.sqf:
```diff
@@ -118,10 +188,42 @@ FO_fn_GUI_updateLabels = {
+        private _displayName = lbText [IDC_FACTORY_WEPS, _i];
+        private _className = _displayName call FO_fn_GUI_findWeaponClassName;
+        private _picturePath = getText (configFile >> "CfgWeapons" >> _className >> "picture");;
+        lbSetPicture [IDC_FACTORY_WEPS, _i, _picturePath];
```
Extra semicolon?

10xJSChad commented on A3A/addons/core/functions/FactoryOverhaul/fn_productionGUIControls.sqf:
```diff
@@ -131,6 +233,9 @@ FO_fn_GUI_addEventHandlers = {
 
         [_factoryID] call FO_fn_GUI_updateLabels;
+
+        [] call FO_fn_GUI_setSelectedWeaponPicture;
```
The line gaps around here are kinda weird.

10xJSChad commented on A3A/addons/core/functions/FactoryOverhaul/fn_productionGUIControls.sqf:
```diff
@@ -46,10 +86,31 @@ FO_fn_GUI_createFactoryTownPairs = {
-		lbAdd [IDC_FACTORY_WEPS, _itemName];
-	} forEach call FO_fn_getUnlockedItems;
+        lbAdd [IDC_CATEGORIES_LIST, _x];
+	} forEach FO_fn_GUI_itemCategories;
```
o jeez o fuck what is this indentation

10xJSChad commented on A3A/addons/core/functions/FactoryOverhaul/fn_productionGUIControls.sqf:
```diff
@@ -46,10 +86,31 @@ FO_fn_GUI_createFactoryTownPairs = {
+        {
+            private _itemDisplayName  = getText (configFile >> "CfgWeapons" >> _x select 0 >> "displayName");
+            private _itemCategory = _itemDisplayName call FO_fn_GUI_displayNameToCategory;
+            //private _wtf = call FO_fn_GUI_displayCategoryToItemCategory;
```
xd

10xJSChad commented on A3A/addons/core/functions/FactoryOverhaul/fn_productionGUIControls.sqf:
```diff
@@ -12,6 +14,44 @@
+
+    private _itemCategory = "";
+    {
+        private _found = [_displayCategory, _x] call BIS_fnc_inString;
```
Why you're using BIS_fnc_inString should probably have been explained in a comment, figuring out how this works was more effort than it should have been.

10xJSChad replied:
Implementation is fine, though, it's just not very clear because one has to read through a lot more code should be necessary to finally figure out exactly what _displayCategory could be, and why we're searching *inside* strings.

But would this not cause problems with non-rhs weapons?
A vanilla weapon's category might be "Rifle", which is going to return true on 
```sqf
["Rifle", "Rhs Rifle"] call BIS_fnc_inString;
```

If this is not the case, let me know why.

NeverendingLearner replied:
Currently, it searching for "AssaultRifle" in "Rhs AssaultRifle". If it were a handgun it would looking for "Handguns" in "Rhs Handguns", and so on, displaying every handgun, Rhs and non.

Honest to god, I abolished the "Rhs" prefix, for more clarity. Later we can add a functionality that actually searches for Rhs, CUP, etc.. weapons only.

NeverendingLearner replied:
One more note, for the current implementation, checking if the strings are equal is enough, don't need to use BIS_fnc_inString. However, I'm gonna keep it there for potential further application.

10xJSChad commented on A3A/addons/core/functions/FactoryOverhaul/fn_productionGUIControls.sqf:
```diff
@@ -46,10 +86,31 @@ FO_fn_GUI_createFactoryTownPairs = {
+    };
+};
+
+
```
FO_fn_GUI_addWeaponsToListbox is very messy, I don't think you need two loops. The code duplication of _itemDisplayName alone is enough to justify not doing that.

This should be refactored into one loop that assigns _itemDisplayName once and handles category checks. It's a bit slower, but significantly better to look at.

NeverendingLearner replied:
Refactored.

10xJSChad commented on A3A/addons/core/functions/FactoryOverhaul/fn_productionGUIControls.sqf:
```diff
@@ -67,20 +128,29 @@ FO_fn_GUI_lookupFactoryIDbyName = {
+FO_fn_GUI_displayNameToCategory = {
+    params ["_selectedItem"];
+    private _className = _selectedItem call FO_fn_GUI_findWeaponClassName;
+    private _itemType = _className call BIS_fnc_itemType;
```
A comment giving an example return value of BIS_fnc_itemType would be nice here, so it's not required to go look it up on the wiki.

NeverendingLearner replied:
Added a comment straight from the wiki

10xJSChad commented on A3A/addons/core/functions/FactoryOverhaul/fn_productionGUIControls.sqf:
```diff
@@ -67,20 +128,29 @@ FO_fn_GUI_lookupFactoryIDbyName = {
+    private _className = "";
+    // Iterate through unlocked items
+    {
+        if (_selectedItem isEqualTo getText (configFile >> "CfgWeapons" >> _x select 0 >> "displayName")) then {
```
```sqf
getText (configFile >> "CfgWeapons" >> _x select 0 >> "displayName")
```
should be a variable, this hurts to look at.

10xJSChad commented:
Say the word and I'll merge.

NeverendingLearner committed:
Revive system modification and Factory Overhaul refactoring

The player is now able to use a one-time self revive kit per life, independent from the Combat Readiness Recovery kit, which should be removed soon.
Also, refactored Factory Overhaul based on review comments.

NeverendingLearner committed:
Update fn_productionGUIControls.sqf

10xJSChad commented on A3A/addons/core/functions/Revive/fn_handleDamage.sqf:
```diff
@@ -42,8 +42,7 @@ if (_part == "" && _damage > 0.1) then
 
-private _makeUnconscious =
-{
+makeUnconscious = {
```
I think this function should have a prefix

10xJSChad commented on A3A/addons/core/functions/Revive/fn_handleDamage.sqf:
```diff
@@ -124,11 +123,27 @@ else
-_damage
+// GAMERARMY: Delays unconsciousnes by 5 seconds, giving the player a brief moment to escape the gun fire.
+// Good for 2-3 people team playthrough.
+FN_DELAYUNCONSCIOUS = {
```
all caps indicates macro, would rename.

10xJSChad commented on A3A/addons/core/functions/Revive/fn_handleDamage.sqf:
```diff
@@ -124,11 +123,27 @@ else
 			{
 				if !(_unit getVariable ["incapacitated",false]) then
```
~~What's going on with this condition? Will incapacitated not always be true since it's set up there at line 47 or am I missing something?~~

That actually seems to not be in the same function, review UI made it look questionable, disregard.

10xJSChad replied:
Also, I don't think going getting downed should have a delay if your one time revive is already used, it looks like you're always gonna have the delay, but the self revive will only work once, this should be reworked to only having the delay if you're gonna be able to self revive, I think.

Although I could be reading it wrong, the formatting is omega fucked on the review page here.

NeverendingLearner replied:
It should be working as intended now.

10xJSChad commented on A3A/addons/core/functions/Revive/fn_handleDamage.sqf:
```diff
@@ -124,11 +123,27 @@ else
+	_unit allowDamage false;
+	_unit setDamage 0.7;
+	_unit allowDamage false;
+	private _duration = time + 5;
```
Any reason for using this over ```sleep```? ```sleep``` should yield so the next script can be scheduled, whereas ```waitUntil```, does not yield and instead just [busy waits](https://en.wikipedia.org/wiki/Busy_waiting) until the condition is met.

If there is no specific reason for using ```waitUntil```, this would be better off as a ```sleep``` so that something else is allowed to run, instead of this script being given execution just to sit there spinning.

10xJSChad commented on A3A/addons/core/functions/Revive/fn_respawn.sqf:
```diff
@@ -17,4 +17,7 @@ if (captive _unit) then {
+_unit setDamage 1;
+
+// GAMERARMY: Reset the one-time use revive.
+_unit setVariable ["oneTimeRevive",false];
```
This should also be reset when using the big orange box at HQ, but that might just call fn_respawn for all I know, I'm not actually sure what it does.

Let me know if this is the case, otherwise try to add it to the box.

10xJSChad reviewed (approved):
Looks mostly good, the busy wait is a bit concerning if it's not required.
Valid for merge, though that should definitely be addressed.

NeverendingLearner committed:
Update Revive System

Added box restore, function prefix, replaced waitUntil with sleep of custom duration, fixed some variable assignments.

10xJSChad commented on A3A/addons/scrt/Common/fn_common_unconsciousEventHandler.sqf:
```diff
@@ -57,6 +57,15 @@ if (reviveKitsEnabled) then {
 				[_nearFriendlyUnits select 0] spawn SCRT_fnc_ai_possessFriendlyUnit;
 			};
 		};
+		case DIK_Z: { // GAMERARMY: One time revive if user presses Z
```
This requires revive kits to be enabled, which isn't entirely desirable behaviour.

10xJSChad replied:
to clarify, this code is present in a 
```c
if (reviveKitsEnabled) {
    ...
}
```

block. I think it'd be best to make it not require revive kits to be enabled, but honestly this doesn't really matter.
It's just good I caught that now before testing it, otherwise I'd probably think the whole thing was bugged since it wouldn't work at all for me.

Just to note though, it being tied to reviveKitsEnabled does seem to mean that all the prompts and stuff around the system like the grace period before being downed will show up even when you can't actually self-revive, since I don't think any other part of this system really checks reviveKitsEnabled.

10xJSChad reviewed (approved):
Looks mostly good, ready for merge if you're fine with it as is. I don't see anything for clearing the one-time revive used flag when using the big orange box at HQ though. That would be something I'd recommend implementing just for convenience.

10xJSChad commented on A3A/addons/core/functions/Revive/fn_handleDamage.sqf:
```diff
@@ -122,13 +128,30 @@ else
 			if (_part in ["head","body"]) then
 			{
-				if !(_unit getVariable ["incapacitated",false]) then
+				if (_unit getVariable ["oneTimeRevive",false] isEqualTo false) then
```
Regarding the check being present twice in fn_makeUnconscious, does this not mean that it could theoretically be triggered twice? Is there a reason for why we don't just have the oneTimeRevive check once at the top or something?

10xJSChad commented:
As a final note, there is some obvious cheese where you can just never use your one-time revive and have the grace period of being able to shoot and move (slowly) after being "downed" if you have someone nearby to revive you every time. I don't know if or how we should handle this.
